src/storage/store/block_engine.rs

```diff
@@ -0,0 +1,549 @@
+        self.trie
+            .exists(ctx, &self.db, sync_id.as_ref())
+            .unwrap_or_else(|err| {
+                error!("Error checking if sync id exists: {:?}", err);
+                false
+            })
+    }
+
+    pub(crate) fn replay_snapchain_txn(
+        &mut self,
+        trie_ctx: &merkle_trie::Context,
+        snapchain_txn: &Transaction,
+        txn_batch: &mut RocksDbTransactionBatch,
+    ) -> Result<Vec<u8>, BlockEngineError> {
+        for message in &snapchain_txn.system_messages {
+            if let Some(ref onchain_event) = message.on_chain_event {
+                match self
+                    .onchain_event_store
+                    .merge_onchain_event(onchain_event.clone(), txn_batch)
+                {
+                    Ok(event) => self
+                        .trie
+                        .update_for_event(trie_ctx, &self.db, &event, txn_batch)?,
+                    Err(err) => {
+                        error!("Unable to merge onchain event: {:#?}", err.to_string())
+                    }
+                }
+            }
+        }
+
+        let account_root =
+            self.trie
+                .get_hash(&self.db, txn_batch, &TrieKey::for_fid(snapchain_txn.fid));
+
+        Ok(account_root)
+    }
+
+    fn generate_block_events(
+        &self,
+        height: Height,
+        timestamp: &FarcasterTime,
+        txn: &mut RocksDbTransactionBatch,
+    ) -> (Vec<BlockEvent>, Vec<u8>) {
+        let mut events = vec![];
+        if let Some(heartbeat_block_interval) = self.heartbeat_block_interval {
+            if height.block_number % heartbeat_block_interval == 0 {
+                let event_seqnum = self.block_event_store.max_seqnum().unwrap() + 1;
+                let data = BlockEventData {
+                    seqnum: event_seqnum,
+                    r#type: BlockEventType::Heartbeat as i32,
+                    block_number: height.block_number,
+                    event_index: events.len() as u64,
+                    block_timestamp: timestamp.to_u64(),
+                    body: Some(block_event_data::Body::HeartbeatEventBody(
+                        HeartbeatEventBody {},
+                    )),
+                };
+                let hash = blake3::hash(data.encode_to_vec().as_slice())
+                    .as_bytes()
+                    .to_vec();
+                let event = BlockEvent {
+                    hash,
+                    data: Some(data),
+                };
+                // Store these events so
+                // (1) It's possible to figuure out the max seqnum easily
+                // (2) It's possible to query over them in an rpc and see what has been produced.
+                self.block_event_store.put_block_event(&event, txn).unwrap();
+                events.push(event);
+            }
+        }
+
+        let events_hash = if events.is_empty() {
+            vec![]
```

Comment:
Hmm, I'm not sure if empty event should hash to empty. What's the value returned by the hasher if you don't call any `update`s on it?